detection/detect_slow_nodes.py

```diff
@@ -229,7 +374,146 @@ def __sortNodesByExecutionTime(self, nodes: list):
+        print()
+        for cluster in sorted(np.unique(np.array(clusters))):
+            representative_label = '(representative)' if cluster == representative_cluster else ''
+            outlier_label = '(outlier)' if cluster_centers[cluster] > threshold else ''
+            center_label = f"(center: {cluster_centers[cluster]:.2f})"
+            print(f" * Cluster {cluster} {representative_label}{outlier_label} {center_label} contains:")
+            cluster_nodes = []
+            node_to_ranks_in_cluster_map = {}
+            for rank, node in self.__rank_to_node_map.items():
+                if rank in cluster_to_ranks[cluster]:
+                    cluster_nodes.append(node)
+                    if node not in node_to_ranks_in_cluster_map:
+                        node_to_ranks_in_cluster_map[node] = []
+                    node_to_ranks_in_cluster_map[node].append(rank)
+            cluster_nodes = np.unique(np.array(cluster_nodes))
+            for node in cluster_nodes:
+                print(f"   | node {node} ({len(node_to_ranks_in_cluster_map[node])})")
+            print()
+
+
+    ###########################################################################
+    ## Outlier detection functions
+
+    def __findClusterOutliers(self, data):
+        """
+        Uses clustering to identify outliers
+        (Currently specialized for rank execution time).
+        """
+        data,                   \
+        clusters,               \
+        cluster_to_times,       \
+        cluster_to_ranks,       \
+        cluster_centers,        \
+        representative_cluster, \
+        representative_center,  \
+        threshold,              \
+        problematic_clusters = self.__clusterTimes(data)
+
+        if len(np.unique(np.array(clusters))) > 1:
+            # identify if representative cluster has slowest center
+            representative_cluster_is_slowest = True
+            slowest_non_representative_center = 0.
+            fastest_non_representative_center = np.inf
+            for cluster_center in {k: v for k, v in cluster_centers.items() if k != representative_cluster}.values():
+                if cluster_center > representative_center:
+                    representative_cluster_is_slowest = False
+                if cluster_center > slowest_non_representative_center:
+                    slowest_non_representative_center = cluster_center
+                if cluster_center < fastest_non_representative_center:
+                    fastest_non_representative_center = cluster_center
+
+            # if representative cluster is slowest, check by how much
+            if representative_cluster_is_slowest:
+                if representative_center - 3 * np.std(cluster_to_times[representative_cluster]) > slowest_non_representative_center:
+                    print()
+                    print(f"     WARNING: Clustering results found most times to be slower than others. No outliers will be detected.")
```

Comment:
I'm thinking that instead of a warning here, we should stop the job from running by not giving the node list with a bunch of slow nodes.